apps/builder/app/services/auth-strategy/ws.server.ts

```diff
@@ -0,0 +1,75 @@
+import type { StrategyVerifyCallback } from "remix-auth";
+import {
+  OAuth2Strategy,
+  type OAuth2Profile,
+  type OAuth2StrategyOptions,
+  type OAuth2StrategyVerifyParams,
+} from "remix-auth-oauth2";
+import { AsyncLocalStorage } from "node:async_hooks";
+
+type DynamicProps = "authorizationEndpoint" | "tokenEndpoint" | "redirectURI";
+
+export type OAuth2StrategyOptionsOverrides = Partial<OAuth2StrategyOptions> &
+  Pick<OAuth2StrategyOptions, DynamicProps>;
+
+const asyncLocalStorage = new AsyncLocalStorage<
+  Partial<OAuth2StrategyOptions>
+>();
+
+/**
+ * The main issue with OAuth2Strategy is that it forces us to define authorizationEndpoint, tokenEndpoint, and redirectURI
+ * at the time of constructing the instance.
+ * However, in our case, these values are dynamic because we use multiple domains, such as development, staging, and production.
+ * The solution is to wrap the options object with a proxy, retrieving the necessary options from localAsyncStorage and proxying
+ * all requests through asyncLocalStorage.run
+ */
+export class WSStrategy<
+  User,
+  Profile extends OAuth2Profile = { provider: string },
+  ExtraParams extends Record<string, unknown> = Record<string, never>,
+> extends OAuth2Strategy<User, Profile, ExtraParams> {
+  constructor(
+    options: OAuth2StrategyOptions,
+    verify: StrategyVerifyCallback<
+      User,
+      OAuth2StrategyVerifyParams<Profile, ExtraParams>
+    >,
+    overrideOptions: (request: Request) => OAuth2StrategyOptionsOverrides
+  ) {
+    super(options, verify);
+
+    this.options = new Proxy<OAuth2StrategyOptions>(this.options, {
+      get: (target, property, receiver) => {
+        const store = asyncLocalStorage.getStore();
+        return (
+          store?.[property as keyof OAuth2StrategyOptions] ??
```

Comment:
shouldn't we check. property in OAuth2StrategyOptions and throw otherwise?

Reply:
const x = {}
x.blabla <- does it throw?

same here

Reply:
yes, but I think you might want a good error message if wrong property is used? just to be nice

Reply:
No just want to be ordinary proxy, some options can be not defined but accessed it’s ok